### PR TITLE
Refactor Take & Fallbacks logic into their own classes

### PR DESCRIPTION
**Pull request by samihibrahim**
For #2

- Extracted Telegram bot Notification Command into its own class
- Extracted Telegram bot Notifications Fallback into its own class
- Extracted Telegram bot Fallback into its own class
- Removed Puzzle

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#21](https://github.com/g4s8/ghman/issues/21) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#21](https://github.com/g4s8/ghman/issues/21) is assigned to @victornoel[/z](https://www.0crat.com/u/victornoel), here is [why](https://www.0crat.com/footprint/CJT4R32Q3/0a65cf9d-392c-4ebc-867d-0d17f3a9c381); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be no monetary reward for this job

File: src/main/java/com/g4s8/ghman/bot/BotApp.java [hunk 1]
```diff
@@ -55,13 +34,11 @@
...
  *
  * @since 1.0
- * @todo #1:30min Refactor this class:
- *  move actual logic to separate takes and keep only
- *  composition structure in this class.
  * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
  * @checkstyle LineLengthCheck (500 lines)
  * @checkstyle ExecutableStatementCountCheck (500 lines)
  */
+@SuppressWarnings("PMD.UnnecessaryFullyQualifiedName")
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/BotApp.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Fallback.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @checkstyle LineLengthCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @samihibrahim why is this needed? Remove it and reformat please

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * @checkstyle ExecutableStatementCountCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Optional;
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Fallback.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import org.cactoos.map.MapEntry;
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Notification Command.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+ * @checkstyle LineLengthCheck (500 lines)
+ * @checkstyle ExecutableStatementCountCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Notification Command.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
+ * @checkstyle LineLengthCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim why is this needed? Remove it and reformat please

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    public TmResponse act(final Update update) throws IOException {
+        final GhUser user = new PgUsers(this.data).user(update.getMessage().getChat()).github();
+        final List<Thread> nts = new Solid<>(user.notifications());
+        return new RsInlineKeyboard(
+            new RsText(new FormattedText("You have %d unread notifications:", nts.size())),
+            new Mapped<>(
+                ntf -> Collections.singleton(
+                    new MapEntry<>(
+                        ntf.subject().getString("title"),
+                        String.format("click:notification#%s", ntf.tid())
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim please use `FormattedText` from cactoos

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.cactoos.text.Joined;
+import org.cactoos.text.TextOf;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Notifications Fallback.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
+ * @checkstyle ClassDataAbstractionCouplingCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+ * @checkstyle LineLengthCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @samihibrahim why is this needed? Remove it and reformat please

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+ * @checkstyle ExecutableStatementCountCheck (500 lines)
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @samihibrahim why is this needed?

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public TmResponse act(final Update update) throws IOException {
+        final GhUser user = new PgUsers(this.data).user(update.getCallbackQuery().getMessage().getChat()).github();
+        final GhThread thread = user.thread(update.getCallbackQuery().getData().split("#")[1]);
+        final Issue issue = new ThreadIssue(user.github(), thread);
+        return new RsText(
+            new FormattedText(
+                "[#%d](%s) - %s\n\n%s",
+                issue.number(),
+                String.format("https://github.com/%s/issues/%d", issue.repo().coordinates(), issue.number()),
```
Comment (victornoel) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @samihibrahim please use `FormattedText` from cactoos

File: src/main/java/com/g4s8/ghman/bot/BotApp.java (hunk 1 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/BotApp.java]: @victornoel this was already present in the original class. I added a todo to this classes to fix both of these exclusions.

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java (hunk 2 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @victornoel  fixed the line length check and added a todo for the ClassDataAbstractionCouplingCheck .

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java (hunk 3 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @victornoel fixed and removed both.

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java (hunk 4 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 5 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 6 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 7 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 8 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @victornoel done.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java (hunk 10 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java (hunk 9 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @victornoel this exclusion is necessary to pass the builds (this class instantiates too many other classes). Added a corresponding exclusion and todo.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java (hunk 11 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @victornoel fixed and removed the exclusion.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java (hunk 12 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsFallback.java]: @victornoel done.

**Comment by samihibrahim** [comment]
@viktornoel removed the unnecessary checkstyle/PMD exclusions & added relevant todos for the ones requiring further refactoring.

**Comment by victornoel** [comment]
@samihibrahim thanks :)

@g4s8 I think it's good now, but maybe check the name of the new classes are coherent with what they do, I haven't thought about it at all

**Comment by samihibrahim** [comment]
@victornoel Thank you for your patience :D

File: src/main/java/com/g4s8/ghman/bot/BotApp.java [hunk 13]
```diff
@@ -111,72 +88,14 @@ public void run() {
...
-                                                    new Mapped<>(
-                                                        Comment.Smart::new,
-                                                        issue.comments().iterate(new Date(thread.lastRead().toEpochMilli()))
-                                                    )
-                                                )
-                                            )
-                                        )
-                                    );
-                                }
+                                    new TkNotificationsFallback(this.data)
```
Comment (g4s8) [inline src/main/java/com/g4s8/ghman/bot/BotApp.java]: @samihibrahim it's not correct class name, this class handles button clicks returned by `/notifications` command and returns single thread from notification, so it's better to rename it to `TkThread` or similar.

File: src/main/java/com/g4s8/ghman/bot/BotApp.java [hunk 14]
```diff
@@ -111,72 +88,14 @@ public void run() {
...
-                                            GhAuthException.class.cast(err).user().uid()
-                                        )
-                                    )
-                                );
-                            } else {
-                                rsp = Optional.empty();
-                            }
-                            return rsp;
-                        }
+                        new FbFallback()
```
Comment (g4s8) [inline src/main/java/com/g4s8/ghman/bot/BotApp.java]: @samihibrahim same here: this class is a fallback for unauthorized access: `FbFallback` → `FbUnauthorized`

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+import com.g4s8.teletakes.rs.TmResponse;
+import java.util.Optional;
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Fallback.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
```
Comment (g4s8) [inline src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @samihibrahim this todo won't be accepted by 0pdd bot, so this PR can not be merged now, you need to extend the message, you can check min words count for pdd puzzle in `.pdd` file in repo root

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+import org.cactoos.list.Solid;
+import org.cactoos.map.MapEntry;
+import org.cactoos.text.FormattedText;
+import org.telegram.telegrambots.api.objects.Update;
+
+/**
+ * Telegram bot Notification Command.
+ *
+ * @since 1.0
+ * @todo #2:30min Implement Unit tests
```
Comment (g4s8) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim same here: pdd text is too short

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+ */
+public final class TkNotificationsCommand implements TmTake {
```
Comment (g4s8) [inline src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @samihibrahim let's name it `TkNotifications` (we can replace its context one level upper without changing class behavior, so it may be called not by command, but by button in the future without changing the implementation)

**Comment by g4s8** [comment]
@samihibrahim also, please do not use force-push command to keep commit history in pull request, right now it's not possible to check what changes was made before review and what changes after

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 17 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @g4s8 Fixed.

File: src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java (hunk 16 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/TkNotificationsCommand.java]: @g4s8 Fixed.

File: src/main/java/com/g4s8/ghman/bot/FbFallback.java (hunk 15 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/FbFallback.java]: @g4s8 Fixed.

File: src/main/java/com/g4s8/ghman/bot/BotApp.java (hunk 14 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/BotApp.java]: @g4s8 Fixed.

File: src/main/java/com/g4s8/ghman/bot/BotApp.java (hunk 13 above)
Comment (samihibrahim) [reply src/main/java/com/g4s8/ghman/bot/BotApp.java]: @g4s8 Fixed.

**Comment by samihibrahim** [comment]
> 
> 
> @samihibrahim also, please do not use force-push command to keep commit history in pull request, right now it's not possible to check what changes was made before review and what changes after

@g4s8 as there is a conflict, should I wait for your approval 1st regarding the above comments? or should I rebase and push now?

**Comment by victornoel** [comment]
> @samihibrahim also, please do not use force-push command to keep commit history in pull request, right now it's not possible to check what changes was made before review and what changes after

Actually you can: in github, the message says `samihibrahim force-pushed the samihibrahim:2 branch from e69de96 to 26bbaf9 4 days ago` and the words `force-pushed` are a link to see the diff of the force push.

**Comment by 0crat** [comment]
@victornoel[/z](https://www.0crat.com/u/victornoel) this job was assigned to you 5days ago. It will be taken away from you soon, unless you close it, see [§8](http://www.zerocracy.com/policy.html#8). Read [this](http://www.yegor256.com/2014/04/13/no-obligations-principle.html) and [this](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html), please.

**Comment by g4s8** [comment]
@samihibrahim sure, you can push as many commits as needed to your local branch for PR

**Comment by g4s8** [comment]
@samihibrahim please resolve merge conflict

**Comment by samihibrahim** [comment]
@g4s8 Conflicts resolved.

**Comment by g4s8** [comment]
@samihibrahim thanks

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ammaratef45[/z](https://www.0crat.com/u/ammaratef45) please review this job completed by @victornoel[/z](https://www.0crat.com/u/victornoel), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#21](https://github.com/g4s8/ghman/issues/21) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ammaratef45** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @victornoel[/z](https://www.0crat.com/u/victornoel)

**Comment by 0crat** [comment]
Quality review completed: +4 point(s) just awarded to @ammaratef45[/z](https://www.0crat.com/u/ammaratef45)
